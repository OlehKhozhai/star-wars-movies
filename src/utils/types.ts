export type Movie = {
  title: string;
  episode_id: number;
  opening_crawl: string;
  director: string;
  producer: string;
  release_date: string;
  created: string;
  url: string;
};

export type FavoriteMovies = Record<string, boolean>;
