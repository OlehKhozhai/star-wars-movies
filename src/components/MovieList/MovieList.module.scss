.root {
  padding: 15px 0;
}

.noMovieFound {
  text-align: center;
}
