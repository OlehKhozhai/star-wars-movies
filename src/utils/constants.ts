export enum APIPath {
  Films = "https://swapi.dev/api/films",
}


export enum LSKey {
  FavoriteMovies = "FAVORITE_MOVIES",
}