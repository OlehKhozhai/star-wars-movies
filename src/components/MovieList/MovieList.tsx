import { memo, FC } from "react";

import MovieItem from "components/MovieItem/MovieItem";
import { FavoriteMovies, Movie } from "utils/types";
import styles from "./MovieList.module.scss";

type MovieListProps = {
  movies: Movie[];
  activeMovie: Movie | null;
  favoriteMoviesConfig: FavoriteMovies;
  onMovieClick: (movie: Movie) => void;
};

const MovieList: FC<MovieListProps> = ({
  movies,
  activeMovie,
  favoriteMoviesConfig,
  onMovieClick,
}) => {
  return (
    <ul className={styles.root}>
      {movies.length === 0 && (
        <h2 className={styles.noMovieFound}>No movies found</h2>
      )}

      {movies.length > 0 &&
        movies.map((movie) => (
          <MovieItem
            key={movie.title}
            {...movie}
            isActive={movie.title === activeMovie?.title}
            isFavorite={favoriteMoviesConfig[movie.title]}
            onMovieClick={onMovieClick}
          />
        ))}
    </ul>
  );
};

export default memo(MovieList);
