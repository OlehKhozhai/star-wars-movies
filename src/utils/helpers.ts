import { LSKey } from "utils/constants";
import { FavoriteMovies, Movie } from "utils/types";

export const getAPIData = <T>(
  url: string,
  params: RequestInit = {}
): Promise<T> =>
  fetch(url, { ...params })
    .then((data) => data.json())
    .then((data) => data)
    .catch((error) => console.log("--error--", error));

export const initializeFavoriteMoviesConfigToLS = (movies: Movie[]): void => {
  const favoriteMovies = localStorage.getItem(LSKey.FavoriteMovies);

  if (!favoriteMovies) {
    const preparedMovies = movies.reduce<FavoriteMovies>(
      (accumulator, movie) => {
        accumulator[movie.title] = false;

        return accumulator;
      },
      {}
    );

    localStorage.setItem(LSKey.FavoriteMovies, JSON.stringify(preparedMovies));
  }
};

export const getFavoriteMoviesConfigFromLS = (): FavoriteMovies => {
  const favoriteMovies = localStorage.getItem(LSKey.FavoriteMovies);

  return favoriteMovies ? JSON.parse(favoriteMovies) : {};
};

export const setFavoriteMoviesConfigToLS = (config: FavoriteMovies): void => {
  localStorage.setItem(LSKey.FavoriteMovies, JSON.stringify(config));
};

export const updateFavoriteMoviesConfig = (
  title: string
): FavoriteMovies => {
  const favoriteMovies = getFavoriteMoviesConfigFromLS();
  const newFavoriteMovies = {
    ...favoriteMovies,
    [title]: !favoriteMovies[title],
  };

  return newFavoriteMovies;
};
