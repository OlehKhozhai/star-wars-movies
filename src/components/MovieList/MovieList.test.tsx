import { create } from "react-test-renderer";

import { movieMock } from "utils/mock";
import { Movie } from "utils/types";
import MovieList from "./MovieList";

jest.mock("./../MovieItem/MovieItem", () => ({
  __esModule: true,
  default: "<---MovieItem--->",
}));

describe("MovieList component", () => {
  it("should render by default", () => {
    const component = create(
      <MovieList
        movies={[]}
        activeMovie={{} as Movie}
        favoriteMoviesConfig={{}}
        onMovieClick={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should render with movie properly", () => {
    const component = create(
      <MovieList
        movies={[movieMock]}
        activeMovie={movieMock}
        favoriteMoviesConfig={{}}
        onMovieClick={() => undefined}
      />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
