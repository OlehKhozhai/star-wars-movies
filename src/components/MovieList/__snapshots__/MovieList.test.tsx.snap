// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList component should render by default 1`] = `
<ul
  className="root"
>
  <h2
    className="noMovieFound"
  >
    No movies found
  </h2>
</ul>
`;

exports[`MovieList component should render with movie properly 1`] = `
<ul
  className="root"
>
  <<---MovieItem--->
    created="2014-12-10T14:23:31.880000Z"
    director="George Lucas"
    episode_id={4}
    isActive={true}
    onMovieClick={[Function]}
    opening_crawl="It is a period of civil war.
Rebel spaceships, striking
from a hidden base, have won
their first victory against
the evil Galactic Empire.

During the battle, Rebel
spies managed to steal secret
plans to the Empire's
ultimate weapon, the DEATH
STAR, an armored space
station with enough power
to destroy an entire planet.

Pursued by the Empire's
sinister agents, Princess
Leia races home aboard her
starship, custodian of the
stolen plans that can save her
people and restore
freedom to the galaxy...."
    producer="Gary Kurtz, Rick McCallum"
    release_date="1977-05-25"
    title="A New Hope"
    url="https://swapi.dev/api/films/1/"
  />
</ul>
`;
